traversal/procedure/GraphProcedure.java

```diff
@@ -54,117 +54,54 @@ public class GraphProcedure implements PermutationProcedure {
 
     private static final Logger LOG = LoggerFactory.getLogger(GraphProcedure.class);
 
-    private final Map<Identifier, ProcedureVertex<?, ?>> vertices;
-    private final ProcedureEdge<?, ?>[] edges;
+    private final ProcedureVertex<?, ?>[] vertices;
     private ProcedureVertex<?, ?> startVertex;
+    private Set<ProcedureVertex<?, ?>> endVertices;
 
-    private GraphProcedure(int edgeSize) {
-        vertices = new HashMap<>();
-        edges = new ProcedureEdge<?, ?>[edgeSize];
+    private GraphProcedure(ProcedureVertex<?, ?>[] vertices) {
+        this.vertices = vertices;
     }
 
     public static GraphProcedure create(GraphPlanner planner) {
-        GraphProcedure procedure = new GraphProcedure(planner.edges().size());
+        GraphProcedure.Builder builder = new Builder();
```

Comment:
we use the `Builder` to register all the vertices against. the `Builder` then contains the transformation logic for converting edges that have an order to vertices that have an order. This helps us not expose a partially correct `GraphProcedure` at any point, and directly produce one from the Builder when the building is finished.